WalletWasabi.Daemon/WasabiAppBuilder.cs

```diff
@@ -0,0 +1,179 @@
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using NBitcoin;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+using WalletWasabi.Services;
+using WalletWasabi.Services.Terminate;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.Daemon;
+
+public class WasabiApplication
+{
+	public WasabiAppBuilder AppConfig { get; }
+	public Global? Global { get; private set; }
+	public Config Config { get; }
+	public SingleInstanceChecker SingleInstanceChecker { get; }
+
+	public WasabiApplication(WasabiAppBuilder wasabiAppBuilder)
+	{
+		AppConfig = wasabiAppBuilder;
+		Config = new Config(LoadOrCreateConfigs(), wasabiAppBuilder.Arguments);
+		SetupLogger();
+		Logger.LogDebug($"Wasabi was started with these argument(s): {(AppConfig.Arguments.Any() ? string.Join(" ", AppConfig.Arguments) : "none")}.");
+		SingleInstanceChecker = new(Config.Network);
+	}
+
+	public async Task<int> RunAsync(Action afterStarting)
+	{
+		if (AppConfig.MustCheckSingleInstance)
+		{
+			var instanceResult = await SingleInstanceChecker.CheckSingleInstanceAsync();
+			if (instanceResult == WasabiInstanceStatus.AnotherInstanceIsRunning)
+			{
+				Logger.LogDebug("Wasabi is already running, signaled the first instance.");
+				return 1;
```

Comment:
As a solution for unnamed variables, you can make constants or even an enum where all possible exit codes are listed. So, when app ends with something different than `0` there's no need to lookup all possible paths to determine where it happens. Plus it helps to avoid return code reusing between different reasons.